packages/netlify-cms-core/src/reducers/entryDraft.js

```diff
@@ -88,13 +88,22 @@ const entryDraftReducer = (state = Map(), action) => {
       });
       return state.set('localBackup', newState);
     }
-    case DRAFT_CHANGE_FIELD:
-      return state.withMutations(state => {
+    case DRAFT_CHANGE_FIELD: {
+      const publishedEntry = action.payload.publishedEntry || Map();
+      const unpublishedEntry = action.payload.unpublishedEntry || Map();
+      let newState = state.withMutations(state => {
         state.setIn(['entry', 'data', action.payload.field], action.payload.value);
         state.mergeDeepIn(['fieldsMetaData'], fromJS(action.payload.metadata));
         state.set('hasChanged', true);
       });
+      const newStateData = newState.getIn(['entry', 'data']);
+
+      (newStateData.equals(unpublishedEntry.get('data')) ||
+        newStateData.equals(publishedEntry.get('data'))) &&
+        (newState = newState.set('hasChanged', false));
 
+      return newState;
```

Comment:
Can this be done with only a single Immutable node?

```
return state.withMutations(state => {
  state.setIn(['entry', 'data', action.payload.field], action.payload.value);
  state.mergeDeepIn(['fieldsMetaData'], fromJS(action.payload.metadata));

  const newData = state.getIn(['entry', 'data']);
  const matchesLastSaved = newData.equals(unpublishedEntry.get('data')) || newData.equals(publishedEntry.get('data'))
 
  state.set('hasChanged', !matchesLastSaved);
});

Reply:
I changed that in my branch https://github.com/netlify/netlify-cms/commit/411d9860fd283b224418b1617a6464fb851e6e1c#diff-9409717c5b67d95c17fad5f0b8165dd8R106